gcs/src/gcs_sm.cpp

```diff
@@ -137,9 +137,54 @@ gcs_sm_destroy (gcs_sm_t* sm)
 {
     gu_mutex_destroy(&sm->lock);
     gu_cond_destroy(&sm->cond);
+    delete sm->pauses_window;
     gu_free (sm);
 }
 
+double
+gcs_sm_paused_in_window_get (gcs_sm_t* const sm, double window_sec,
+                             bool erase_outdated)
+{
+    assert(sm->fc_auto_evict);
+    long long now = gu_time_monotonic();
+    long long window_ns = now - window_sec * 1000000000LL;
+
+    if (gu_unlikely(gu_mutex_lock (&sm->lock))) abort();
+
+    std::vector<long long> & window = *sm->pauses_window;
+    auto size = window.size();
+    assert(size % 2 == sm->pause);
+
+    long long total_ns = 0;
+    ssize_t i;
+    // paused => handle the last pause separately, start from the next pair
+    if (sm->pause) {
+        assert(window[size-1] == sm->stats.pause_start);
+        total_ns += now - sm->stats.pause_start;
+        i = size - 2;
+    // not paused => num pauses == num conts, just start from the end
+    } else {
+        i = size - 1;
+    }
+    assert(i % 2); // there are pairs of pause and cont moments
+
+    for (; i > 0 && window[i] > window_ns; i -= 2) {
+        long long cont = window[i];
+        long long pause = std::max(window[i-1], window_ns);
+        total_ns += cont - pause;
+    }
+
+    /* Remove outdated moments. We need to move all fresh elements, but
+       it seems good anyway, because window probably won't be big. */
+    if (erase_outdated && i > 0) {
+        window.erase(window.begin(), window.begin()+i+1);
+        assert(window.size() % 2 == sm->pause);
+    }
+
+    gu_mutex_unlock (&sm->lock);
+    return total_ns * 1.0e-9;
```

Comment:
Similarly to `window_sec`, it looks like it is possible to avoid frequent multiplications.

Reply:
Done